[TG-2755] Fix nondet-static option for static initializers

majakusber opened the pull request:
If `nondet-static` option is used, this makes sure that static variables that have a static initializer are not set to the value in the static initiliazer but nondet-initialized (by nondet-initializing all static variables of the class in `clinit_wrapper`).

Commit:
Record nondet-static option in java_bytecode_language

Commit:
Pass information to clinit_wrapper constructors

Pass parameters necessary to nondet-initialize variables

majakusber commented:
Tests and test-gen bump will follow.

smowton commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -157,6 +157,11 @@ void java_bytecode_languaget::get_language_options(const cmdlinet &cmd)
 
+  if(cmd.isset("nondet-static"))
+  {
+    nondet_static = true;
```
`generate_nondet_static_initializers`?

majakusber replied:
There are two reasons for this name - first, it copies the name of the flag itself which seems to be often the case in this function (it makes it easier to see the connection then?) and second, this option does not only generate nondet staic initializers. It nondet initializes everything in CPROVER_init and also goes over static initializers and uses nondet there.

smowton commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -184,9 +184,12 @@ gen_clinit_eqexpr(const exprt &expr, const clinit_statest state)
 /// \param class_name: name of the class to generate clinit wrapper calls for
 /// \param [out] init_body: appended with calls to clinit wrapper
```
Update docs

smowton commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -384,7 +421,10 @@ static void create_clinit_wrapper_symbols(
 /// \return the body of the static initialiser wrapper
```
Update docs

smowton commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -205,9 +208,43 @@ static void clinit_wrapper_do_recursive_calls(
+
+          parameters.max_nonnull_tree_depth =
+            is_non_null_library_global(symbol.first)
+              ? object_factory_parameters.max_nonnull_tree_depth + 1
```
Should this just be `std::max(1, object_factory_parameters.max_nonnull_tree_depth)`?

smowton reviewed (approved):
Looks good, some nitpickery

peterschrammel reviewed (commented):
@svorenova, there's a failed test: https://travis-ci.org/diffblue/cbmc/jobs/410887870#L7400

Commit:
Nondet-initialize variables in clinit_wrapper constructor

If nondet-static option is used

majakusber commented:
@peterschrammel The test should be passing now.

majakusber commented:
Pointer bump and regression tests can be found here: https://github.com/diffblue/test-gen/pull/2128

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 7ff9a8a3394eaf313ea1c98a5be84110ffdd5898).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80782147
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

peterschrammel reviewed (approved):
It would be good to have at least one JBMC test as well.

majakusber commented:
@peterschrammel It seems `nondet-static` is not an option in jbmc. Is there a particular reason for it? If not then I can add it.

peterschrammel commented:
>  can add it.

Yes, please.

Commit:
Add nondet-static option to JBMC

The option already appears in process_goto_functions before, the flag just needed to be read.

Commit:
Add a regression test for nondet-static

Commit:
Mark final fields in Java as constants

This will also make sure that they won't get nondet-initialized with
nondet-static option

majakusber commented:
@smowton @peterschrammel I added the nondet-static option to JBMC and a test for it. I also changed the solution a bit - see the clinit_wrapper cont. commit - to make sure that `static final` fields get handled correctly.

smowton commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -209,6 +220,43 @@ static void clinit_wrapper_do_recursive_calls(
+                  size_t(1), object_factory_parameters.max_nonnull_tree_depth)
+              : object_factory_parameters.max_nonnull_tree_depth;
+
+          gen_nondet_init(
```
Careful -- gen-nondet-init mutates the symbol table, which invalidates iterators over the symbol table. You must first pick which identifiers to mess with, then retrieve and run gen-nondet-init for each in turn.

majakusber replied:
Done :)

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2376b329fa3526f82317d16563d76cc74d01310c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81038517

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8762296d09f2eb8205c3d91a3a98cc912c4f7f40).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81038249

majakusber commented:
Squashed some commits, waiting for the final round of CI.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 86f51f54bf3097746fcabc0b2c92b94cf3bd5567).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81123619

majakusber commented:
Extended the regression test with various combinations of public/private/final fields, initialization with a function call, using models, fields in inner classes...

Commit:
Perform nondet-init in clinit_wrapper even if clinit doesn't exist

This just moves the whole if(nondet-static) block outside of the
if(find_sym_it != symbol_table.symbols.end()) block and updates the
comment

Commit:
Add a comment to static field with the name of the field

majakusber commented:
Test-gen pointer bump: https://github.com/diffblue/test-gen/pull/2195

majakusber commented:
@smowton @peterschrammel Can you please have a quick second look at the last 4 commits? The tests have been extended, and I added the last three small commits related to this.

smowton commented:
Looks good
